feat(postgres sink): Add postgres sink

jorgehermo9 opened the pull request:
Closes #15765

This PR is not 100% ready by my side and there will likely be a few things wrong, but had a few questions and wanted to know if the direction seems right... So I would like an initial round of review if possible.

I tested the sink and it seems to be working, but I lack a lot of knowledge about Vector's internals and I'm not sure if the implementation is okay. 

I inspired a lot from the `databend` and `clickhouse` sinks, but left a few questions as TODOs in the source. I found this sink a bit different from the others, as the others had the `request_builder` thing and encoding the payload in bytes (as most of the sinks are http based).. But I didn't think that fitted well in this case, as in the `sqlx` API I should wrap the events with the `sqlx::types::Json` type and that will do all the encoding with `serde` internally. 
 

If someone want to manually test it, I used this Vector config:

```toml
[sources.demo_logs]
type = "demo_logs"
format = "apache_common"

[transforms.payload]
type = "remap"
inputs = ["demo_logs"]
source = """
.payload = .
"""

[sinks.postgres]
type = "postgres"
inputs = ["payload"]
endpoint = "postgres://postgres:postgres@localhost/test"
table = "test"
```

Run postgres server with `podman run -e POSTGRES_PASSWORD=postgres -p 5432:5432 docker.io/postgres`

and execute the following with `psql -h localhost -U postgres`:
```sql
CREATE DATABASE test;
```
then execute `\c test`
and last:
```sql
CREATE TABLE test (message TEXT, timestamp TIMESTAMP WITH TIME ZONE, payload JSONB);
```

And then, you will see logs in that table:

![image](https://github.com/user-attachments/assets/acc85d5e-4561-440c-b495-2e511b7b842b)

jorgehermo9 commented on scripts/integration/postgres/test.yaml:
```diff
@@ -1,5 +1,6 @@
 features:
 - postgresql_metrics-integration-tests
+- postgres-integration-tests
```
this name it is not so consistent with the `postgresql_metrics-integration-tests` feature, I'm open to suggestions here

jorgehermo9 commented on src/sinks/postgres/config.rs:
```diff
@@ -0,0 +1,147 @@
+}
+
+/// Configuration for the `postgres` sink.
+#[configurable_component(sink("postgres", "Deliver log data to a PostgreSQL database."))]
```
should I call this sink `postgres` or `postgres_logs`?

jorgehermo9 replied:
I'm thinking if this can evolve to integrate with other postgres flavours such as timescaledb https://docs.timescale.com/use-timescale/latest/metrics-logging/

jorgehermo9 replied:
Added support for metrics & traces in https://github.com/vectordotdev/vector/pull/21248/commits/19796274e230a28daf2c18c29b69e1bae15baf6d, resolving this conversation!

jorgehermo9 commented on src/sinks/postgres/service.rs:
```diff
@@ -0,0 +1,140 @@
+            // TODO: If a single item of the batch fails, the whole batch will fail its insert.
+            // Is this intended behaviour?
+            sqlx::query(&format!(
+                "INSERT INTO {table} SELECT * FROM jsonb_populate_recordset(NULL::{table}, $1)"
```
the table configuration can be a victim of sql injection, but in my opinion, we shouldn't avoid that kind of attacks at this level and the user should be responsible of ensuring that there is not sql injection in the config... The `databend` sink works like this

pront replied:
Hm, I suppose `sqlx` does not support parameterized table names? Does the query builder help here? If none of the above works, then we can leave as is.

jorgehermo9 replied:
I don't think that could help in this case. See [this statement](https://docs.rs/sqlx/latest/sqlx/struct.QueryBuilder.html#warning-beware-sql-injection-vulnerabilities-and-untrusted-input) about `sqlx`'s query builder. 

And we cannot use a variable bind (`$` syntax) in postgres for table names, as the prepared statements are bounded to a query plan and it cannot change if the target table changes.

I think this is the better way to do it... `sqlx` does not check for sql injection

pront replied:
Got it. Is it just the `table` config param currently, so we can add a validation check when building the config verification. This becomes more complicated if `table` becomes a template, in that case we should just add a notice in the docs to communicate that this sink isn't trying to be smart about security.

jorgehermo9 replied:
> we could just add a notice in the docs to communicate that this sink isn't trying to be smart about security.

Yes, I think we should do this at least for now... Clickhouse,databend and other sinks do not try to protect users from slq injection though.

pront replied:
Sounds good. Feel free to resolve when the docs are updated.

jorgehermo9 replied:
Done on 3e7ff2335616679d86d52728f5ec7fe0614c4a8a

jorgehermo9 commented on src/sinks/postgres/config.rs:
```diff
@@ -0,0 +1,144 @@
+    pub endpoint: String,
+
+    /// The table that data is inserted into.
+    pub table: String,
```
Should we make the table templatable? Like the `clickhouse`  sink. That would complicate the code a little bit (with `KeyPartitioner` and so. If yes, I would like some guidance about it if possible

pront replied:
It is a nice feature but not a must-have, we can do this incrementally. Once we finalized the rest of the comments we can come back to this if you are motivated to add this feature.

jorgehermo9 commented on src/sinks/postgres/integration_tests.rs:
```diff
@@ -0,0 +1,111 @@
+}
+
+#[tokio::test]
+async fn test_postgres_sink() {
```
I think just a single test is too little.. But I couldn't figure anything else to test. This test is very similar of the integration test from `databend` sink

pront replied:
There are some more interesting things we can do here, at the very least send more than one events. Also, we could test failures such as sending a badly formatted payload.

jorgehermo9 replied:
Okay, going to include more tests!

aliciascott reviewed (approved):
good for docs

jorgehermo9 commented on src/sinks/postgres/integration_tests.rs:
```diff
@@ -0,0 +1,111 @@
+use std::future::ready;
+use vector_lib::event::{BatchNotifier, BatchStatus, BatchStatusReceiver, Event, LogEvent};
+
+fn pg_host() -> String {
```
Copied those utility functions from the `posgres_metrics` source integration tests

pront replied:
Let's do some small refactoring here. 
* The folder `src/test_util` seems like a good location to put these utils.
* You can use `#[cfg(any(feature = "postgresql_metrics-integration-tests", feature = "postgres_sink-integration-tests"))]` in `src/test_util/mod.rs`.

pront commented on src/sinks/postgres/config.rs:
```diff
@@ -0,0 +1,144 @@
+    /// The table that data is inserted into.
+    pub table: String,
+
+    /// The postgres connection pool size.
```
Here it would be useful to explain what this pool is used for. Maybe a link to relevant docs would suffice.

jorgehermo9 replied:
done in 21f3ad3a56090b94cb8f58ec7bf3bfbf4fb51439. Do you think it is enough?

I also have doubts about using a connection pool. Can the event batches be executed in parallel for this sink? I don't know the specifics of vector's internals... https://github.com/vectordotdev/vector/blob/21f3ad3a56090b94cb8f58ec7bf3bfbf4fb51439/src/sinks/postgres/sink.rs#L24

If the batches of events can be processed in parallel, then a connection pool is beneficial. If the batches are processed sequentially, then we should use a single postgres connection as a pooled connection does not have sense

pront reviewed (commented):
Hi @jorgehermo9, thank you for this sizable contribution! On a high level, it looks great. I did a first review and left some comments. Don't hesitate to follow up, happy to discuss details.

jorgehermo9 commented:
Thank you very much for the review @pront! I'm kinda busy these days but I will revisit this as soon as I can :)

pront commented:
There are a few failing checks. Also, let's add a new `postgres` semantic scope in https://github.com/vectordotdev/vector/blob/master/.github/workflows/semantic.yml. I will review once these are addressed. Thank you!

jorgehermo9 commented:
Thanks @pront. Will work on this PR these days, I'll ping you whenever it is ready for another round. Thank you so much!

jorgehermo9 commented:
Hi @pront. I'm sorry for the delay. I added a bunch of new integration tests and I'm ready for another review round.

I still have a few doubts left maked as `// TODO`  in code. 

I also wonder if I should enable metrics & traces inputs in this shink (see this thread https://github.com/vectordotdev/vector/pull/21248#discussion_r1855446211). If so, I will add a few more tests similar to `insert_multiple_events` test I added in this PR. 

I think there are a few opened comment threads, feel free to resolve them if you think my answer is enough. Didn't resolve them myself just in case my answer is not clear.

jorgehermo9 commented on src/sinks/postgres/integration_tests.rs:
```diff
@@ -0,0 +1,305 @@
+
+const POSTGRES_SINK_TAGS: [&str; 2] = ["endpoint", "protocol"];
+
+fn pg_url() -> String {
```
Simplified this a little bit. 

I see that https://github.com/vectordotdev/vector/pull/21248#discussion_r1801866517 stated that this should be moved to `test_utils.rs`, but as this is simplified, I don't know if its worth. 

And please not that now `PG_URL` env var is required. Before, it was optional and fallback to localhost if not set.

Take a look to the `postgresql_metrics.rs` diff 
![image](https://github.com/user-attachments/assets/720b3983-734f-4865-8f06-0478ea97acaa)

I prefer it to be simplified and not do any type of magic fallback... But I'm open to change it to what you suggest of course.

pront replied:
Nit: Hmm, I am bit skeptical of changing `src/sources/postgresql_metrics.rs` behavior  in this PR, even it's just for tests. For example, when I grep for `pg_socket_url()`, I don't see it used in multiple places. It's fine to have a different `pg_url()` in this file.

jorgehermo9 replied:
Understood, I will revert `postgresql_metrics.rs` entirely

pront commented:
> Hi @pront. I'm sorry for the delay. I added a bunch of new integration tests and I'm ready for another review round.
> 
> I still have a few doubts left maked as `// TODO` in code.

Hi @jorgehermo9, I will take another look now. Thanks!
 
> I also wonder if I should enable metrics & traces inputs in this shink (see this thread [#21248 (comment)](https://github.com/vectordotdev/vector/pull/21248#discussion_r1855446211)). If so, I will add a few more tests similar to `insert_multiple_events` test I added in this PR.

Yes, I was going to suggest exactly this. Add more inputs and observe how it works. 

> Metrics can be useful for sinks like timescale #939. But I don't know if the current postgres sink is compatible with timescaledb. I think with timescaledb we should use this same sink but with other flag (see [this comment](https://github.com/vectordotdev/vector/issues/21308#issuecomment-2380873820), so I think we should really think ahead if this sink will be used for more than just logs.

Are we talking about [timescaledb](https://github.com/timescale/timescaledb) and [risingwave](https://github.com/risingwavelabs/risingwave)? Are there more? I am not familiar with all the nuances and I don't think we can plan for all extensions in advance. The proposed `flavour` (or `flavor`) property can be added incrementally without breaking the behavior of the sink.

pront commented:
Just dropping by to say that I am looking forward to introducing this new integration in Vector. And don't hesitate to give me a shout if you need anything.

pront commented:
Of course! No need to apologize at all 😅 

I just wanted to express my excitement about this work and to let you know I’m here to support whenever you’re ready.

jorgehermo9 commented on src/test_util/integration.rs:
as required in https://github.com/vectordotdev/vector/pull/21248#discussion_r1801866517

Is it ok to move those functions to `src/test_util/integration.rs`? The original discussion suggested bout moving it to `src/test_util/mod.rs`, but I think this module is a better fit for integration-test-only utils

pront replied:
> but I think this module is a better fit for integration-test-only utils

Yes, good point.

jorgehermo9 commented on src/test_util/mod.rs:
```diff
@@ -67,6 +67,9 @@ pub mod mock;
 
 pub mod stats;
 
+#[cfg(test)]
```
Is this cfg gate okay? I don't see any feature flag related to  "any integration test". I would like to compile this module only if *any* (and no all) of the integration tests feature flags behind the feature flag `all-integration-tests` are specified..

pront replied:
We can technically write....

```
#[cfg(any(
    test,
    feature = "amqp-integration-tests",
    feature = "appsignal-integration-tests",
    feature = "aws-integration-tests",
    feature = "axiom-integration-tests",
    feature = "azure-integration-tests",
    feature = "chronicle-integration-tests",
    feature = "clickhouse-integration-tests",
    feature = "databend-integration-tests",
    feature = "datadog-agent-integration-tests",
    feature = "datadog-logs-integration-tests",
    feature = "datadog-metrics-integration-tests",
    feature = "datadog-traces-integration-tests",
    feature = "dnstap-integration-tests",
    feature = "docker-logs-integration-tests",
    feature = "es-integration-tests",
    feature = "eventstoredb_metrics-integration-tests",
    feature = "fluent-integration-tests",
    feature = "gcp-cloud-storage-integration-tests",
    feature = "gcp-integration-tests",
    feature = "gcp-pubsub-integration-tests",
    feature = "greptimedb-integration-tests",
    feature = "http-client-integration-tests",
    feature = "humio-integration-tests",
    feature = "influxdb-integration-tests",
    feature = "kafka-integration-tests",
    feature = "logstash-integration-tests",
    feature = "loki-integration-tests",
    feature = "mongodb_metrics-integration-tests",
    feature = "nats-integration-tests",
    feature = "nginx-integration-tests",
    feature = "opentelemetry-integration-tests",
    feature = "postgresql_metrics-integration-tests",
    feature = "prometheus-integration-tests",
    feature = "pulsar-integration-tests",
    feature = "redis-integration-tests",
    feature = "splunk-integration-tests",
    feature = "webhdfs-integration-tests"
))]
```

I don't know if there's a more elegant way to achieve the above.

jorgehermo9 replied:
yeah :smile: I fear that we will forgot about maintaining that list synchronized with the IT test feature flags in Cargo.toml... If you prefer to specify that, I'm okay to include it in this PR, but as there is no much code in that module as of right now... maybe is it okay to just leave `#[cfg(test)]`?

pront replied:
We can leave this as is. Resolving.

jorgehermo9 commented:
> Are we talking about [timescaledb](https://github.com/timescale/timescaledb) and [risingwave](https://github.com/risingwavelabs/risingwave)? Are there more? I am not familiar with all the nuances and I don't think we can plan for all extensions in advance. The proposed flavour (or flavor) property can be added incrementally without breaking the behavior of the sink.

Yes, I was talking about those two specifically. I agree to not plan in advance. I will test how this sink behaves with metrics and traces and enable it in regular postgres if its works ok. Will update about that soon (I think this is one of the main things I have left from your review, correct me if I'm wrong)

pront commented:
> Hi @pront, I have more time available these days and will work in this PR. I have left to test how this sink behaves with metrics and traces and enable if its works ok. Will update about that soon. I think I answered all of your comments, but correct me if I'm wrong

🎉 Awesome, I will prioritize reviewing this PR.

jorgehermo9 commented:
I tried with metrics (I have left to test with traces)

with this config
```toml
[sources.internal_metrics]
type = "internal_metrics"

[sinks.postgres]
type = "postgres"
inputs = ["internal_metrics"]
endpoint = "postgres://postgres:password123@localhost/test"
table = "metrics"
```

and this commands to spin-up a local postgres and connect to it:
```console
docker run -e POSTGRES_PASSWORD=password123 -p 5432:5432 docker.io/postgres
psql -h localhost -U postgres
```
and this sql statements inside `psql`:
```sql
CREATE DATABASE test;
\c test
CREATE TABLE metrics(name TEXT, namespace TEXT, tags JSONB, timestamp TIMESTAMP WITH TIME ZONE, kind TEXT, counter JSONB, gauge JSONB, aggregated_histogram JSONB);
```

The table is populated with the given metrics and we can query it:
![image](https://github.com/user-attachments/assets/662ad708-15d1-4cc9-8cb7-f8d996348751)


I had to create a database with the union of the fields of all metrics (counter JSONB, gauge JSONB, aggregated_histogram JSONB) but once the `table` parameter is templatable, we could have different postgres tables per metric type (one for counters, other for gauges, etc)  and that would be very instersting!

this looks very promising ;)


I will update the comments once I do more tests!

jorgehermo9 commented:
@pront I don't know how to propertly generate traces in Vector.. Is there any minimal configuration example to get traces similar to what I did with metrics? 

I can try to ingest traces within integration tests, but I'd like to do a manual testing as I did with metrics & logs

jorgehermo9 commented on src/sinks/postgres/integration_tests.rs:
```diff
@@ -0,0 +1,355 @@
+}
+
+#[derive(Debug, Serialize, Deserialize, FromRow)]
+struct TestCounterMetric {
```
I had to create this struct to derive `FromRow` so we can read from the `sqlx` query result. Didn't find any other alternative... 

I tried to deserialize the `sqlx` result into a `serde_json::Value` and then `let metric: Metric = serde_json::from_value(value)` but it seems that `serde_json::Value` does not implement `sqlx::FromRow`...
This is the only workaround that I could get working.

pront replied:
This is test code, so it's fine to do it this way 👍

jorgehermo9 commented:
Added support for metrics and traces in 19796274e230a28daf2c18c29b69e1bae15baf6d. Left a few `TODO`s for the review and Im ready for another review round!

jorgehermo9 commented on src/sinks/postgres/integration_tests.rs:
```diff
@@ -0,0 +1,355 @@
+}
+
+#[tokio::test]
+async fn insert_metric() {
```
I just added this one test for metrics ingestion. Do you think it is enough or should we think of more test cases?

I just used one type of metric (absolute counter), because I didn't want to bloat these tests... Should I test with more types (histogram, gauges...) or not?

pront replied:
Ideally, yes we would have tests for each type. But due to the size of this PR and the fact that the sink will be advertised as beta, we can leave as is. If you feel inspired you can follow up later to improve the test suite.

jorgehermo9 commented on src/sinks/postgres/service.rs:
```diff
@@ -0,0 +1,154 @@
+    type Response = PostgresResponse;
+
+    fn is_retriable_error(&self, _error: &Self::Error) -> bool {
+        // TODO: Implement this
```
I don't know what type of errors should consider Vector as retryable... Query timeout? 

`sqlx::Error` does have an `Io` variant https://docs.rs/sqlx/latest/sqlx/enum.Error.html#variant.Io and a pool timeout variant https://docs.rs/sqlx/latest/sqlx/enum.Error.html#variant.PoolTimedOut, maybe we can retry in those two cases exclusively?

pront replied:
I can imagine this can get complicated... We should start with a very small set and the community can followup and make the case for adding more types here. So, yes we can start with these two!

jorgehermo9 replied:
Perfect, done in 91c9b0255! Resolving this

jorgehermo9 replied:
Terminating the DB once the sink is connected:

![image](https://github.com/user-attachments/assets/1427609d-c88a-4775-aebb-09264364973c)

jorgehermo9 commented on src/test_util/mod.rs:
```diff
@@ -231,6 +234,10 @@ pub fn temp_dir() -> PathBuf {
     path.join(dir_name)
 }
 
+pub fn temp_table() -> String {
```
Thoughts about this naming?

jorgehermo9 replied:
Done in 702f7ad9c

jorgehermo9 replied:
done in 702f7ad9c

jorgehermo9 commented on src/sinks/postgres/integration_tests.rs:
```diff
@@ -0,0 +1,483 @@
+        .unwrap();
+    let create_table_sql = format!(
+        "CREATE TABLE {table} (agent_version TEXT, env TEXT, error_tps BIGINT, host TEXT,
+        language TEXT, spans trace_span[], target_tps BIGINT, trace_id BIGINT)"
```
I typed the span (`TestTraceSpan` struct) instead of clearing a `jsonb[]` column. I wanted to test insertions in postgres composite types (https://www.postgresql.org/docs/current/rowtypes.html).

Is is okay to leave the test of this feature (inserting in composite types) in the traces tests, or should I create another test to explicitly test this...? For example, it would be a test similar to `insert_single_event`, but with field that is a composite type (instead of `payload: serde_json::Value`) in a very similar way to what is done here with `trace_span` postgres type.

I think it is okay to just test the composite type inserting here... But maybe we should separate concerns and use here `spans JSONB[]` and create a simpler test for composite type inserting?

I don't know if I'm explaining well :/

pront replied:
For integration tests, it's okay to leave the composite type insertion test within the trace tests. Since integration tests validate the entire system's behavior (including how traces interact with the database), testing composite type insertion within traces is acceptable. I suppose composite types will be used beyond just traces, but that's fine.

jorgehermo9 commented on src/sinks/postgres/integration_tests.rs:
```diff
@@ -0,0 +1,483 @@
+
+#[derive(Debug, Serialize, Deserialize, sqlx::Type, FromRow)]
+#[sqlx(type_name = "trace_span")]
+struct TestTraceSpan {
```
Created this type because of https://github.com/vectordotdev/vector/pull/21248/files#r1948104371

Resolve this conversation once read!

jorgehermo9 commented:
Derived from https://github.com/vectordotdev/vector/pull/21248/commits/e9eb6889b228ad337810534fa2511c79258785cc I manually tested composite type insertion

with this config
```toml
[sources.demo_logs]
type = "demo_logs"
format = "apache_common"


[transforms.structured]
type = "remap"
inputs = ["demo_logs"]
source = """
. = {
    "message": .message,
    "structured_field": {
        "string": "value",
        "integer": 1,
        "float": 1.1,
        "boolean": true,
    }
}
"""

[sinks.postgres]
type = "postgres"
inputs = ["structured"]
endpoint = "postgres://postgres:password123@localhost/test"
table = "structured_table"
```

and those sql statements:
```sql
CREATE TYPE structured AS (string TEXT, integer BIGINT, float real, boolean BOOLEAN);
CREATE TABLE structured_table(message text, structured_field structured);
```

we can insert pg's composite types (records) https://www.postgresql.org/docs/current/rowtypes.html from nested fields, instead of having to use `JSONB` for those fields

I would like to document also an example of this in the `website` docs! This looks great :smile:

pront commented on src/sinks/postgres/integration_tests.rs:
```diff
@@ -0,0 +1,483 @@
+        ("resource".into(), Value::from(resource)),
+        ("type".into(), Value::from("a_type")),
+        ("trace_id".into(), Value::Integer(123)),
+        ("span_id".into(), Value::Integer(456)),
+        ("parent_id".into(), Value::Integer(789)),
+        ("start".into(), Value::from(timestamp())),
+        ("duration".into(), Value::Integer(1_000_000)),
+        ("error".into(), Value::Integer(404)),
+        (
+            "meta".into(),
+            Value::Object(ObjectMap::from([
+                ("foo".into(), Value::from("bar")),
+                ("bar".into(), Value::from("baz")),
+            ])),
+        ),
+        (
+            "metrics".into(),
+            Value::Object(ObjectMap::from([
+                ("a_metric".into(), Value::Float(NotNan::new(0.577).unwrap())),
+                ("_top_level".into(), Value::Float(NotNan::new(1.0).unwrap())),
+            ])),
+        ),
+    ])
+}
+
+pub fn create_trace(resource: &str) -> TraceEvent {
+    let mut t = TraceEvent::default();
+    t.insert(event_path!("language"), "a_language");
+    t.insert(event_path!("agent_version"), "1.23456");
+    t.insert(event_path!("host"), "a_host");
+    t.insert(event_path!("env"), "an_env");
+    t.insert(event_path!("trace_id"), Value::Integer(123));
+    t.insert(event_path!("target_tps"), Value::Integer(10));
+    t.insert(event_path!("error_tps"), Value::Integer(5));
+    t.insert(
+        event_path!("spans"),
+        Value::Array(vec![Value::from(create_span(resource))]),
+    );
+    t
+}
```
In this instance we duplicate them. Nit: I don't think we need all the fields apart from `trace_id` and `spans`. This will also help simplify test structs.

jorgehermo9 replied:
Agree!

Done in 04c8f39b9. I only removed fields from the trace and not from the spans. If you think some span fields can be removed too, I'll do it, but i wasn't sure about this. Leaving this unresolved, but please feel free to resolve it if span fields are ok

pront commented on src/sinks/postgres/integration_tests.rs:
```diff
@@ -0,0 +1,504 @@
+// it is like inserting null values explicitly, it does not use table's default values.
+// https://dba.stackexchange.com/questions/308114/use-default-value-instead-of-inserted-null
+// https://stackoverflow.com/questions/49992531/postgresql-insert-a-null-convert-to-default
+// TODO: this cannot be fixed without a workaround involving a trigger creation, which is beyond
```
👍 Leaving this comment as a reminder to address.

jorgehermo9 replied:
Where can I document this (In which file I mean)? I only see per-setting documentation and not documentation about the whole sink in the documentation generated from source (`postgres.cue`)
![image](https://github.com/user-attachments/assets/92e3f7a1-1a0d-4621-8c25-18708cbbeb80)

pront replied:
After you introduce a new postgres.cue file (not the generated `.../base/`) you can add a how it works section. You can copy paste the format from another sink like `exclusive_route`.

pront commented on src/sinks/postgres/service.rs:
```diff
@@ -0,0 +1,154 @@
+        let finalizers = events.take_finalizers();
+        let metadata_builder = RequestMetadataBuilder::from_events(&events);
+        let events_size = NonZeroUsize::new(events.estimated_json_encoded_size_of().get())
+            .expect("payload should never be zero length");
```
Hmm, panicking here seems too draconian. One alternative would be to `warn!` (throttled) and just not sent a request.

jorgehermo9 replied:
Yeah, you're right :laughing: 

Inspired from a few other places in the codebase https://github.com/search?q=repo%3Avectordotdev%2Fvector%20%20%20%20%20%20%20%20%20%20%20%20%20.expect(%22payload%20should%20never%20be%20zero%20length%22)%3B&type=code

Although, I changed it to a `TryFrom` in 13bd5bb2a. Resolve this conversation if the TryFrom approach is more idiomatic now!

pront commented on src/sinks/postgres/service.rs:
```diff
@@ -0,0 +1,154 @@
+        let metadata_builder = RequestMetadataBuilder::from_events(&events);
+        let events_size = NonZeroUsize::new(events.estimated_json_encoded_size_of().get())
+            .expect("payload should never be zero length");
+        // TODO: is this metadata creation correct?
```
Yes,
```suggestion
```

pront commented on src/sinks/postgres/service.rs:
```diff
@@ -0,0 +1,154 @@
+
+impl Service<PostgresRequest> for PostgresService {
+    type Response = PostgresResponse;
+    // TODO: previosly, we had here `slqx::PostgresError`, but as
```
Sounds reasonable:

```
use snafu::Snafu;

#[derive(Debug, Snafu)]
pub enum PostgresServiceError {
    #[snafu(display("Database error: {source}"))]
    PostgresError { source: sqlx::Error },

    #[snafu(display("Serialization error: {source}"))]
    VectorCommonError { source: vector_common::Error },
}
```

jorgehermo9 replied:
Yep, that is exactly what I thought!

jorgehermo9 replied:
Done in 91c9b0255. Resolving!

pront reviewed (approved):
Thank you @jorgehermo9, this is starting to look awesome. I did another round. I think we are almost there. Tried to answer to all TODOs. But I will do another review after your future updates.

pront commented on src/sinks/postgres/service.rs:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    fn events_sent(&self) -> &GroupedCountByteSize {
+        // TODO: Is this correct?
```
This is correct but `events_sent` is a misleading name. We should refactor, preferably in another PR since this is not related to postgres.

jorgehermo9 replied:
Thanks for the explanation!

jorgehermo9 commented:
Ready for another round, thanks for the review! Left unresolved a few threads as I was unsure, feel free to resolve them if the changes look ok!

pront commented:
Hi @jorgehermo9, I am looking at this now. Two quick things that popped out:
* The spell checker is failing
* You can add a new scope in `.github/workflows/semantic.yml`

pront commented:
This 99% ready to go. We will need some more docs. See [this comment](https://github.com/vectordotdev/vector/pull/22072#issuecomment-2576203234) from another PR.

pront commented on src/sinks/postgres/config.rs:
```diff
@@ -0,0 +1,161 @@
+        Ok((VectorSink::from_event_streamsink(sink), healthcheck))
+    }
+
+    // TODO: allow for Input::all()
```
We support all types now.
```suggestion
```

jorgehermo9 replied:
forgot to remove, sorry!

pront commented on src/sinks/postgres/config.rs:
```diff
@@ -0,0 +1,161 @@
+#[typetag::serde(name = "postgres")]
+impl SinkConfig for PostgresConfig {
+    async fn build(&self, _cx: SinkContext) -> crate::Result<(VectorSink, Healthcheck)> {
+        // TODO: make connection pool configurable. Or should we just have one connection per sink?
```
We can start with the simplest method and later we can expose configuration for this.

jorgehermo9 replied:
As It is already configurable (forgot to remove TODO once I implemented before), do you mind leaving it as-is right now? I could remove it if you prefer. Resolving this comment and removing comment thought

pront commented on src/sinks/postgres/sink.rs:
```diff
@@ -0,0 +1,49 @@
+        input
+            // TODO: is this batch setting ok?
+            .batched(self.batch_settings.as_byte_size_config())
+            // TODO: use request builder?
```
You could see an example like `src/sinks/databend/sink.rs`. This works though.

pront commented on src/sinks/postgres/sink.rs:
```diff
@@ -0,0 +1,49 @@
+
+    async fn run_inner(self: Box<Self>, input: BoxStream<'_, Event>) -> Result<(), ()> {
+        input
+            // TODO: is this batch setting ok?
```
```
A batcher config using the `ByteSizeOf` trait to determine batch sizes.
```
So yes, remove TODO.

pront commented on src/sinks/postgres/service.rs:
```diff
@@ -0,0 +1,169 @@
+                .collect::<Result<Vec<_>, _>>()
+                .context(VectorCommonSnafu)?;
+
+            // TODO: If a single item of the batch fails, the whole batch will fail its insert.
```
I think it makes sense for this to be transactional.

pront reviewed (approved):
Tried to answer some questions that were left in TODO section. Ideally we should remove all TODOs (expect ones like "// TODO add tests for more metric types")

aliciascott reviewed (approved):
good for docs

jorgehermo9 commented:
@pront I just noticed this tutorial about acknowledgements https://github.com/vectordotdev/vector/blob/master/docs/tutorials/sinks/1_basic_sink.md#acknowledgements. I didn't explicitly call `finalizers.update_status(EventStatus::Delivered)` in this PR. Is this called automatically with the `DriverResponse` trait impl? https://github.com/vectordotdev/vector/blob/24021e61db1097c363614407fa4c143220f5261f/src/sinks/postgres/service.rs#L103

Also, when an error is returned from the service https://github.com/vectordotdev/vector/blob/24021e61db1097c363614407fa4c143220f5261f/src/sinks/postgres/service.rs#L127, will `EventStatus::Errored` be called automatically?

I am 90% sure of this behaviour and that ACKs works with my implementation, but wanted to raise this just in case..

jorgehermo9 commented:
Added an initial documentation @pront, I still have pending to add this note https://github.com/jorgehermo9/vector/blob/feature/postgres-sink/src/sinks/postgres/integration_tests.rs#L339

expect to finish all by end of day, but what is done right now can be reviewed though.

sorry for the delay!!

jorgehermo9 commented on website/cue/reference/services/postgres.cue:
```diff
@@ -0,0 +1,10 @@
+	url:      urls.postgresql
+	versions: null
+
+	description: "[PostgreSQL](\(urls.postgresql)) is a powerful, open source object-relational database system that uses and extends the SQL language combined with many features that safely store and scale the most complicated data workloads."
```
Postgres? PostgreSQL? the sink is called `postgres`, but in description and other text sections I called it `PostgreSQL`... It is a bit inconsistent

jorgehermo9 commented on website/cue/reference/components/sinks/postgres.cue:
```diff
@@ -0,0 +1,223 @@
+	support: {
+		requirements: []
+		warnings: []
+		notices: [
```
notices are currently not rendered in website. Opened https://github.com/vectordotdev/vector/issues/22592

pront replied:
Let's move to `warnings` for now.

jorgehermo9 commented on website/cue/reference/components/sinks/postgres.cue:
```diff
@@ -0,0 +1,223 @@
+		stateful:      false
+	}
+
+	features: {
```
I'm not sure about everything under `features`. Copied it from other cue files, but I don't really now if it is okay

jorgehermo9 commented on website/cue/reference/components/sinks/postgres.cue:
```diff
@@ -0,0 +1,223 @@
+			}
+			compression: enabled: false
+			encoding: enabled:    false
+			tls: enabled:         false
```
I think theoretically we could enable ssl directly in connection strings https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING-URIS with `sslmode=require`, and I think it would simply load certificates from the machine.

I didn't set `tls: enabled: true` here as `TlsConfig`(https://github.com/vectordotdev/vector/blob/1dcc0df4e303994d7ff6ea686fc42040cd1224c9/lib/vector-core/src/tls/settings.rs#L362) is not used in the sink and we can't tweak tls config at sink level (path for ca file and etc) as we do in [other sinks](https://github.com/vectordotdev/vector/blob/master/src/sinks/http/config.rs#L91).

I think we could leave it as-is and implement tls tweaking it in future PRs

jorgehermo9 commented on website/cue/reference/components/sinks/postgres.cue:
```diff
@@ -0,0 +1,223 @@
+			compression: enabled: false
+			encoding: enabled:    false
+			tls: enabled:         false
+			to: {
```
How is this `to` section used? I couldn't find it rendered in the website

pront replied:
I wonder if this is another case of an unrendered field.

jorgehermo9 commented on website/cue/reference/components/sinks/postgres.cue:
```diff
@@ -0,0 +1,223 @@
+				max_bytes:    10_000_000
+				timeout_secs: 1.0
+			}
+			request: {
```
is this for `https://github.com/vectordotdev/vector/blob/a0205c9259725253b388b88ddedb9bc92c5d217c/src/sinks/postgres/config.rs#L67` ?

pront replied:
Yes, also see `website/cue/reference/components/sinks.cue`

jorgehermo9 replied:
Oh, I see! Thanks :smile:

jorgehermo9 commented on website/cue/reference/components/sinks/postgres.cue:
```diff
@@ -0,0 +1,223 @@
+
+	classes: {
+		commonly_used: false
+		delivery:      "at_least_once"
```
not sure about `at_least_once`, but I think.. it is?

jorgehermo9 commented:
Can't update license files because of 

```console
>cargo vdev build licenses
Error: Could not read "/home/jorge/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/allocator-api2-0.2.16/license"

Caused by:
    Is a directory (os error 21)
Error: command: cd "/home/jorge/github/vector-fork" && "dd-rust-license-tool" "write"
  failed with exit code: 1
```

When listing that directory,
```
> tree license
license
├── APACHE
└── MIT

1 directory, 2 files
```

It seems that those two licenses are contained inside. 

Checking `Cargo.lock`, this is displayed:
```toml
[[package]]
name = "allocator-api2"
version = "0.2.16"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0942ffc6dcaadf03badf6e6a2d0228460359d5e34b57ccdc720b7382dfbd5ec5"
```

It seems that `allocator-api2@0.2.16`  is being used. Browsing its repository at that point, https://github.com/zakarumych/allocator-api2/tree/735f821db310410dc975ba316388eaf6d9751b13/license It indeed is a directory. But, in never versions such as `allocator-api@0.2.21` https://github.com/zakarumych/allocator-api2/tree/63cd7fcc2f8854b5821c7054d026e8a4647acde1 they got rid of that directory and split the license files into two `LICENSE-APACHE` and `LICENSE-MIT` license files at the root repo.


Given this, I think the only two solutions would be to either update `dd-rust-license-tool` to work with directories, or to update `allocator-api` to `v0.2.21` using a cargo `patch`, as it is a transitive dependency.

pront commented:
> Can't update license files because of
> 
> ```
> > cargo vdev build licenses
> Error: Could not read "/home/jorge/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/allocator-api2-0.2.16/license"
> 
> Caused by:
>     Is a directory (os error 21)
> Error: command: cd "/home/jorge/github/vector-fork" && "dd-rust-license-tool" "write"
>   failed with exit code: 1
> ```
> 
> When listing that directory,
> 
> ```
> > tree license
> license
> ├── APACHE
> └── MIT
> 
> 1 directory, 2 files
> ```
> 
> It seems that those two licenses are contained inside.
> 
> Checking `Cargo.lock`, this is displayed:
> 
> ```toml
> [[package]]
> name = "allocator-api2"
> version = "0.2.16"
> source = "registry+https://github.com/rust-lang/crates.io-index"
> checksum = "0942ffc6dcaadf03badf6e6a2d0228460359d5e34b57ccdc720b7382dfbd5ec5"
> ```
> 
> It seems that `allocator-api2@0.2.16` is being used. Browsing its repository at that point, https://github.com/zakarumych/allocator-api2/tree/735f821db310410dc975ba316388eaf6d9751b13/license It indeed is a directory. But, in never versions such as `allocator-api@0.2.21` https://github.com/zakarumych/allocator-api2/tree/63cd7fcc2f8854b5821c7054d026e8a4647acde1 they got rid of that directory and split the license files into two `LICENSE-APACHE` and `LICENSE-MIT` license files at the root repo.
> 
> Given this, I think the only two solutions would be to either update `dd-rust-license-tool` to work with directories, or to bump `allocator-api` to `v0.2.21` using a cargo `patch`, as it is a transitive dependency.

I think I fixed it with https://github.com/vectordotdev/vector/pull/21248/commits/e358befd2bcae148a193796709aeaf3a1531ca14

pront reviewed (approved):
Seems this is 99% ready. Some final tweaks to the docs and spell checking fixes are required.

jorgehermo9 commented:
> I think I fixed it with https://github.com/vectordotdev/vector/commit/e358befd2bcae148a193796709aeaf3a1531ca14

Just for curiosity sake, how did you do it? I ran `cargo vdev build licenses` and didn't work :open_mouth:

jorgehermo9 commented:
Oh, I see, raw `dd-rust-license-tool write` works... I wonder if we broke `cargo vdev build licenses` for everyone with this... :disappointed:

jorgehermo9 commented:
> spell checking fixes are required.

I think spellcheck was fixed in https://github.com/vectordotdev/vector/pull/21248/commits/627dbba40545a73046f3080372d057f79dcc61c4

pront commented:
> cargo vdev build licenses

You can try ` cargo vdev check licenses` and also we have the CI

jorgehermo9 commented:
Addressed all comments in 8a204693aa029ba4ac15fc18792a4a3c82642984. I had to update the website layout to handle `exactly_once` delivery, as this was the first sink of that kind!

pront commented:
Exciting, thank you @jorgehermo9 🎉

jorgehermo9 commented:
Thank you very much for your great help here @pront! I know this have been a long one and I'm very grateful for your patience! 🎉🎉

jmaupetit commented:
Thanks for this amazing contribution @jorgehermo9 :heart: 

@pront In which release do you plan to ship it?

jorgehermo9 commented:
@jmaupetit  it will land on the next release which is scheduled on monday https://calendar.google.com/calendar/u/0/embed?src=c_qsq1nrsgetsspkn7pjrqh6cclc@group.calendar.google.com&ctz=America/New_York&pli=1

jmaupetit commented:
> @jmaupetit it will land on the next release which is scheduled on monday https://calendar.google.com/calendar/u/0/embed?src=c_qsq1nrsgetsspkn7pjrqh6cclc@group.calendar.google.com&ctz=America/New_York&pli=1

Oh wow, that was quick, thanks a lot! :muscle:
